dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/converter/TrackerBundleParamsConverter.java

```diff
@@ -0,0 +1,159 @@
+ *            |
+ *            |_ EVENT 3
+ *            |_ EVENT 4
+ * </pre>
+ *
+ * This converter will transform the object into:
+ *
+ * <pre>
+ *
+ * TrackerBundleParams
+ *   |
+ *   __TEI
+ *  |__ENROLLMENT 1, ENROLLMENT 2
+ *  |
+ *  |__EVENT 1, EVENT 2, EVENT 3, EVENT 4
+ *
+ * </pre>
+ *
+ *
+ * @author Luciano Fiandesio
+ */
+public class TrackerBundleParamsConverter
+    extends
+    StdConverter<TrackerBundleParams, TrackerBundleParams>
+{
+    @Override
+    public TrackerBundleParams convert( TrackerBundleParams bundle )
+    {
+        if ( hasNestedStructure( bundle ) )
+        {
+            // collect a list of TEI ids
+            List<String> teiIds = getTrackedEntityIds( bundle );
+            List<String> enrollmentIds = new ArrayList<>();
+
+            // collect all the enrollments of the TEIs
+            List<Enrollment> enrollments = bundle.getTrackedEntities().stream()
+                .flatMap( l -> l.getEnrollments().stream() ).collect( Collectors.toList() );
+
+            List<Event> events = new ArrayList<>();
+
+            for ( Enrollment enrollment : enrollments )
+            {
+                if ( teiIds.contains( enrollment.getTrackedEntityInstance() ) )
+                {
+                    bundle.getEnrollments().add( enrollment );
+                    // accumulate all enrollment ids
+                    enrollmentIds.add( enrollment.getEnrollment() );
+                    // accumulate all events
+                    events.addAll( enrollment.getEvents() );
+                }
+                else
+                {
+                    failOrSkip( "Enrollment", enrollment.getEnrollment(), "Tracked Entity Instance", bundle.getAtomicMode() );
+                }
+            }
+
+            for ( Event event : events )
+            {
+                if ( enrollmentIds.contains( event.getEnrollment() ) )
+                {
+                    bundle.getEvents().add( event );
+                }
+                else
+                {
+                    failOrSkip( "Event", event.getEvent(), "Enrollment", bundle.getAtomicMode() );
+                }
+            }
+
+            // remove the nested structure
+            for ( TrackedEntity tei : bundle.getTrackedEntities() )
+            {
+                tei.setEnrollments( Collections.emptyList()  );
+            }
+        }
+
+        return bundle;
+    }
+
+    private List<String> getTrackedEntityIds( TrackerBundleParams bundle )
+    {
+        return bundle.getTrackedEntities().stream().map( TrackedEntity::getTrackedEntity )
+            .collect( Collectors.toList() );
+    }
+
+    private boolean hasNestedStructure( TrackerBundleParams bundle )
+    {
+        return bundle.getEnrollments().isEmpty();
+    }
+
+    private void failOrSkip( String type, String uid, String parent, AtomicMode atomicMode )
+    {
+        if ( atomicMode.equals( AtomicMode.ALL ) )
+        {
+            throw new ImportException(
```

Comment:
I wonder if this maybe would be better done in the validation stages that @netroms is working on, so that we can check for invalid items and remove / report them there... it would probably play better with the rest of the importer

Reply:
Yes, responsibility-wise, it would make more sense to move the validation logic in the validation stage but this code is executed at deserialization stage (see the `@JsonDeserialize(converter = TrackerBundleParamsConverter.class)` on top of the `TrackerBundleParams` object) so, this kind of validation has to be done here - unless  we flatten the data regardless of any parent child relationship

Reply:
Validation has been removed from this component and delegated to the validation layer